### DISCUSSION TITLE
Value Error: buffer source array is read-only

### DISCUSSION THREAD

**Discussion by KMayank29** [explosion/spaCy#12216] [Help: Coding & Implementations] [answered]
![Screenshot from 2023-02-02 00-17-57](https://user-images.githubusercontent.com/36352999/216135116-aa7b1e39-b326-46b7-b799-9f619abbcd40.png)
![Screenshot from 2023-02-02 00-19-18](https://user-images.githubusercontent.com/36352999/216135395-6c88f180-e0a6-4b89-8d8c-fd030f83c037.png)

I have attached the code snippet and the error I am having. Could you help me why spacy model is throwing the error when sent as ray remote task/actor?
Thanks

**Comment by polm** [comment, marked as answer]
Please share code or terminal output as text, not as screenshots. 

I'm not familiar enough with ray to say exactly what's going on, but some searching suggests this is an issue with differences in how Cython and ray handle memory. We haven't updated it in a while and it was designed to be used in a different way, but [spacy-ray](https://github.com/explosion/spacy-ray) might be a helpful reference for you.

**Reply by KMayank29** [reply]
Thanks for the response @polm 
When I run the same code on GPU using `en_core_web_trf` there is no error but as soon as I use cnn based models (md,lg) it throws error `buffer source array is read-only`.
The entite traceback:
`output = model.ops.gather_add(vectors, keys)
  File "thinc/backends/numpy_ops.pyx", line 440, in thinc.backends.numpy_ops.NumpyOps.gather_add
  File "stringsource", line 660, in View.MemoryView.memoryview_cwrapper
  File "stringsource", line 350, in View.MemoryView.memoryview.__cinit__
ValueError: buffer source array is read-only`

**Reply by polm** [reply]
Sorry, but as I mentioned, we don't have enough familiarity with Ray to help you with this. It doesn't seem to be an issue with spaCy specifically, but rather something to do with memory management in Cython. I think you would have more luck asking for help through one of [the official Ray channels](https://github.com/ray-project/ray#getting-involved). 

If you do figure out what the issue is and there's something we can do in spaCy to improve it, we'd be glad to integrate a fix.
